docs/internal/generics.md

```diff
@@ -156,3 +156,110 @@ let z : U32 & imm = obj.convert(y); // ERROR: Not able to define what type y is
+
+Example:
+```ts
+class Generic[A] { field : A & imm; ... };
+...
+// Here, `A` is `U32`
+let foo : Generic[U32] = ...;
+
+// Here, `A` is `MyClass`
+let bar : Generic[MyClass] = ...;
+```
+
+In the example above, there is one polymorphic class, `Generic[A]` where `A` is the polymorphic type parameter, and two concrete classes (specialisations): `Generic[U32]` and `Generic[MyClass]`.
+
+The compiler generates the required specialisations and replaces each definition with its respective concrete classes, then removes the generic class.
+Since generic classes can only be used by specialisation, all other (potentially infinite) possibilities of types parameters don't exist in the program, so the generic class is elided and we're left with only the specialisations.
+
+The end result would be:
+```ts
+// Definition of monomorphic classes only
+class Generic_U32 { field : U32 & imm; ... };
+class Generic_MyClass { field : MyClass & imm; ... }
+
+// Direct use of the concrete classes
+let foo : Generic_U32 = ...;
+let bar : Generic_MyClass = ...;
+```
+
+Note that type unions (`A | B`) are still monomorphic types.
+To resolve which implementation works with each inner type (`A` or `B`) one must use `match` inside the method.
+
+Example:
+```ts
+class Generic[A] {
+  field : A & imm;
+  // Implement constructor with match on types
+  add(other : A & imm) {
+    match (A)
+    { Numeric => field + other }
+    { String => field.append(other) }
+    { _ => Throw[None] } // Or some better error handling
+  }
+}
+...
+// This creates a monomorphic union type
+let foo : Generic[U32 | String] = ...;
+
+// This creates a monomorphic single type
+let bar : Generic[U32] = ...;
+
+// This union of Numeric types isn't quite static
+let baz : Generic[U32 | F64] = ...;
+```
+
+On the first case, `foo` is a monomorphic union between `U32` and `String`, which means, at run-time, either can be passed to `add` and we can't decide at compile time.
+So the implementation retains the `match` and will check if the types are `Numeric` or `String`.
+Note that both `field` and `other` must have the same type (both are `A`).
+
+However, on the second case, `bar` is a compile-time single monomorphic type, which guarantees that it will always implement the `Numeric` interface and therefore there's only one `match` statement that fits.
+In this case, the compiler is free to elide the `match` altogether and emit only the case where the type is `Numeric`, ie. `field + other`.
+In this case, the type is guaranteed to be `U32` to the call to `+` is a static call to `U32::+(U32, U32)`.
+
+Finally, in the last case, `baz`'s types are both `Numeric`, but the `+` operator for each is different.
+So, while the compiler is free to elide all other `match` cases, the dispatch of the `+` method is dynamic, ie. `field.+(other)`.
```

Comment:
I don't think this type checks.  `U32::+(U32 & imm) : U32 & imm` and `F64::+(F64 & imm) : F64 & imm` do not both satisfy the same interface.  The field is allowed to be a `U32` and the argument an `F64` (or vice versa) and the addition would then fail.

Reply:
What about `Self` types?

```ts
interface Numeric {
  +(Self & imm) : Self & imm;
};
```

Wouldn't `Self` be `U32` in one case and `F64` in the other, thus both satisfying the interface?

Reply:
That's a fine definition for the interface, but it isn't used by the call in your `add` because it takes only a single argument.  The receiver is a concrete type that is one out of `U32` and `F64`.  The other argument is *also* a concrete type that is one out of `U32` and `F64` but there is no requirement that it be the same one.  Your `Numeric` should be defined something like this:

```verona
interface Numeric
{
  +(Self & imm, Self & imm) : Self & imm;
}
```

But this doesn't work because `Self` is a concrete type that is either `U32` or `F64` but the second argument is `U32 | F64`.  You could make `add` work by writing this:

```verona
interface Numeric[T]
{
  +(Self & imm, T & imm) : Self & imm;
}
```

But then neither `U32` nor `F64` would satisfy `Numeric[U32 | F64` because U32 neither implements a method that matches:

```verona
+(Self & imm, (U32 | F64) & imm) : Self & imm;
```

Reply:
Ah, good point!

Furthermore, in the original generic implementation, the `Numeric` match won't catch `U32 | F64` either, so it would be throw.

What if we constrain that further:

```ts
    match (A)
    { Float => field + other }
    { Integer => field + other }
    { String => field.append(other) }
    { _ => Throw[None] } // Or some better error handling
```

Because one of the union types is `Integer` and the other is `Float`, there would only be one candidate for each dispatch, where `typeof(field) == typeof(other)`.

I'm not advocating to write that kind of code, which would fail if the union type was `U32 | U64`, but would that be a possible compiler transformation (to elide other cases and only lower the one match)?

Reply:
Marking this a type error with an explanation why.